components/region_cache_memory_engine/src/range_manager.rs

```diff
@@ -0,0 +1,260 @@
+// k3-k5, and k5-k10.
+// k1-k3 and k5-k10 will be new ranges inserted in self.ranges with meta dervied
+// from meta of k1-k10 (only safe_ts and can_read will be derived). k1-k10 will
+// be removed from self.ranges and inserted to self.historical_ranges. Then,
+// k3-k5 will be in the self.evicted_ranges. Now, we cannot remove the data of
+// k3-k5 as there may be some snapshot of k1-k10. After these snapshot are
+// dropped, k3-k5 can be acutally removed.
+#[derive(Default)]
+pub struct RangeManager {
+    // Each new range will increment it by one.
+    id_allocator: IdAllocator,
+    // Range before an eviction. It is recorded due to some undropped snapshot, which block the
+    // evicted range deleting the relevant data.
+    historical_ranges: BTreeMap<CacheRange, RangeMeta>,
+    evicted_ranges: BTreeSet<CacheRange>,
+    // ranges that are cached now
+    ranges: BTreeMap<CacheRange, RangeMeta>,
+}
+
+impl RangeManager {
+    pub(crate) fn ranges(&self) -> &BTreeMap<CacheRange, RangeMeta> {
+        &self.ranges
+    }
+
+    pub(crate) fn new_range(&mut self, range: CacheRange) {
+        let range_meta = RangeMeta::new(self.id_allocator.allocate_id());
+        self.ranges.insert(range, range_meta);
+    }
+
+    pub fn set_range_readable(&mut self, range: &CacheRange, set_readable: bool) {
+        if let Some(meta) = self.ranges.get_mut(range) {
+            meta.can_read = true;
+        }
+    }
+
+    pub fn set_safe_ts(&mut self, range: &CacheRange, safe_ts: u64) -> bool {
+        if let Some(meta) = self.ranges.get_mut(range) {
+            if meta.safe_ts > safe_ts {
+                return false;
+            }
+            meta.safe_ts = safe_ts;
+            true
+        } else {
+            false
+        }
+    }
+
+    pub(crate) fn overlap_with_range(&self, range: &CacheRange) -> bool {
+        self.ranges.keys().any(|r| r.overlaps(range))
+    }
+
+    // Acquire a snapshot of the `range` with `read_ts`. If the range is not
+    // accessable, None will be returned. Otherwise, the range id will be returned.
+    pub(crate) fn range_snapshot(&mut self, range: &CacheRange, read_ts: u64) -> Option<u64> {
+        let Some(range_key) = self.ranges.keys().find(|&r| r.contains(range)).cloned() else {
+            return None;
+        };
+        let meta = self.ranges.get_mut(&range_key).unwrap();
+
+        if read_ts <= meta.safe_ts || meta.evicted || !meta.can_read {
+            // todo(SpadeA): add metrics for it
+            return None;
+        }
+
+        meta.range_snapshot_list.new_snapshot(read_ts);
+        Some(meta.id)
+    }
+
+    // If the snapshot is the last one in the snapshot list of one cache range in
+    // historical_ranges, it means one or some evicted_ranges may be ready to be
+    // removed physically.
+    // So, here, we return a vector of ranges to denote the ranges that are ready to
+    // be removed.
+    pub(crate) fn remove_range_snapshot(
```

Comment:
Since remove_range_snapshot is called typically by the unified read pool thread, which is latency critical. 
We should not return a range list to be deleted here as it's not good to block user thread for deleting the data.